CRAYSAT-1530: Document `sat bootprep` improvements (#29)
Cray-HPE/docs-sat: docs/usage.md

These sections may be specified in any order, and any of the sections may be
omitted if desired.

#### Bootprep input file schema version

The `sat bootprep` input file is validated against a versioned schema
definition. The input file should specify the version of the schema with which
it is compatible under a `schema_version` key. For example:

```yaml
---
schema_version: 1.0.2
```

To get the current `sat bootprep` input file schema version understood by the
version of `sat` installed on the system, look at the value of the `version` key
in the schema definition:

```screen

> ncn-m001# sat bootprep view-schema | grep -E '^version:'

Author: Yeah, I guess not.

version: '1.0.2'
```

The `sat bootprep run` command will validate the schema version specified in the
input file and make sure it is compatible with the schema version understood by
the current version of `sat bootprep`. See the description of the

> `schema_version` property in the JSON Schema file for more details about schema

Author: Yeah, I'll just drop JSON Schema here. I think I'll say, "See the description of the `schema_version` property in the bootprep input file schema for more details about schema version validation."

Then I'll refer to the sections "Viewing the exact schema specification" and "Generating user-friendly documentation".

These sections may be specified in any order, and any of the sections may be
omitted if desired.

> #### Bootprep input file schema version

Author: Yeah, I was considering that as I also didn't like the unnecessary 5 levels of headings. I'll make that change.

> #### Bootprep input file schema version

Author: Do you think it could even be shortened to "Providing a schema version"? It's a subsection of "Editing a bootprep input file" already, so I think the context is fairly clear.

> #### Bootprep input file schema version

Reviewer: Yes, I think that's a great idea!

The `sat bootprep` input file is validated against a versioned schema
definition. The input file should specify the version of the schema with which
it is compatible under a `schema_version` key. For example:

```yaml
---
schema_version: 1.0.2
```

> To get the current `sat bootprep` input file schema version understood by the
> version of `sat` installed on the system, look at the value of the `version` key

Author: This suggestion feels less accurate to what I'm trying to say. What if we just simplify it to:

```suggestion
The current `sat bootprep` input file schema version can be viewed with the
following command:
```

I'm really just trying to convey how to get the version of the schema that is currently understood by the version of `sat bootprep` installed on the system, so that the user can use this schema version in their input files. I'm also working on providing the schema version in the bootprep input files generated by `sat bootprep generate-example` in [CRAYSAT-1545](https://jira-pro.its.hpecorp.net:8443/browse/CRAYSAT-1545), which should make this easier for the end user.

> version of `sat` installed on the system, look at the value of the `version` key

Reviewer: I like this change, but I think you'll also want to delete line 51, then.

supplied, but if it is absent, the version is assumed to be the latest version
found in the cray-product-catalog.

Alternatively, a configuration layer may be specified by explicitly referencing
the desired configuration repository, along with the branch containing the
intended version of the Ansible playbooks. A commit hash may be specified by

> replacing `branch` with `commit`.

Author: Applied this one manually due to required changes in surrounding context and will push a commit momentarily.

> ```screen

Author: Yeah, the ellipses get highlighted weirdly in that case, IIRC from Jack's original PR for these instructions. Having said that, it might be more helpful to just show an entire CFS configuration here so it's easier to see that this is a layer? I think I'll do that.

> ```screen

Author: It was a little more complicated than that because the following paragraph assumes that these two layers were in a single configuration. I ended up combining these two snippets into one YAML snippet.

- An `image_ref` key can be used to specify another image from the input file
  using its `ref_name`.

Images may also contain a `configuration` key specifying a CFS configuration to
use to customize the built image. If a configuration is specified, then
configuration groups must also be specified using the
`configuration_group_names` key.

Images may also contain a `ref_name` key specifying a unique name that can be
used to refer to this image within the input file in other images or in session
templates. The `ref_name` key allows references to images from the input file
that have dynamically generated names as described in [Dynamic variable
substitutions](#dynamic-variable-substitutions).

Images may also contain a `description` key. It is not currently used, but it
may be useful to describe the image in the bootprep input file.

Here is an example of an image using an existing IMS recipe as its base. This
example will build an IMS image from that recipe. It will then configure it with
a CFS configuration named `example-compute-config`. The `example-compute-config`
CFS configuration can be defined under the `configurations` key in the same

> input file, or it can be a CFS configuration that already exists in CFS. Running

Author: Yeah, it's a bit of a mouthful. I'll say "an existing CFS configuration" instead.

`sat bootprep` against this input file will result in an image named
`example-compute-image`.

```yaml
images:
- name: example-compute-image
  description: >
    An example compute node image built from an existing IMS recipe.
  base:
    ims:
      name: example-compute-image-recipe
      type: recipe
  configuration: example-compute-config
  configuration_group_names:
  - Compute
```

> Here is an example showing the definition of two images. The first image is

Author: Maybe I'm missing something, but I'm actually not showing the dynamic variable substitution here. I did that intentionally to try to gradually introduce topics.

> Here is an example showing the definition of two images. The first image is

Reviewer 2: Nope, that was my fault. I was confusing the image references with dynamic variable substitution.

The HPC CSM Software Recipe provides a manifest defining the versions of each
HPC software product included in the recipe. These product versions can be used
in the `sat bootprep` input file using Jinja2 templating syntax.

##### Selecting the HPC CSM Software Recipe version

By default, the `sat bootprep` command will use the product versions from the
latest installed version of the HPC CSM Software Recipe. However, this can be
overridden with the `--recipe-version` command-line argument to `sat bootprep
run`.

For example, to explicitly select the `22.11.0` version of the HPC CSM Software
Recipe, specify `--recipe-version 22.11.0`:

```screen
ncn-m001# sat bootprep run --recipe-version 22.11.0 compute-and-uan-bootprep.yaml
```

##### Values supporting Jinja2 templating

Jinja2 templating of the `sat bootprep` input file is performed individually
for each supported property. That is, the entire bootprep input file is not
rendered by the Jinja2 template engine.

The following values support rendering as a Jinja2 template:

- The `name` of each configuration under the `configurations` key.
- The values of the following keys of each layer under the `layers` key in a
  configuration:

>     - `name`

Reviewer 2: @haasken-hpe and I talked offline about this and decided we'll allow the MD030 and MD007 violations for now, while I continue to work on adding the lint configuration to docs-sat. In both of these cases we think there's a good case for configuring the lint rules to prefer these styles.

>     - `name`

Author: I discussed this with Eli a bit in slack. Both MD007 and MD030 are configurable. I don't feel too strongly about 2 spaces vs. 4 spaces for MD007. MD030 was an intentional choice that we made when working with Ben. We wanted a consistent visual indent for all multi-line list items, and we felt that 3 spaces was a little annoying to maintain in an editor, e.g. when the tab stop was set to 2 or 4 spaces. We'll discuss those two items in scrum and decide as part of Eli's CRAYSAT-1571 work.

I will address MD012 here.

- `node_groups`: HSM groups to add to the boot set
- `rootfs_provider`: the root file system provider
- `rootfs_provider_passthrough`: parameters to add to the `rootfs=` kernel
  parameter

As mentioned above, the parameters under `bos_parameters` are passed through
directly to BOS. Refer to the section **BOS Session Templates** of the [*Cray

> System Management Documentation*](https://cray-hpe.github.io/docs-csm/) for more

Author: This was actually intentionally not a link to the specific section because we don't want to have to maintain the `en-13` part of the link. There is some discussion about this (in the other direction, from docs-csm to docs-sat) here: https://github.com/Cray-HPE/docs-csm/pull/2366#discussion_r966317466

Jessica says it's the approach techpubs is taking for now for linking between different products' documentation.

  configuration: example-configuration
  bos_parameters:
    boot_sets:
      example_boot_set:
        kernel_parameters: ip=dhcp quiet
        node_roles_groups:
        - Compute
        rootfs_provider: cpss3
        rootfs_provider_passthrough: dvs:api-gw-service-nmn.local:300:nmn0
```

Here is an example of a BOS session template that refers to an image from the
input file by its `ref_name`. This requires that an image defined in the input
file specifies `example-image` as the value of its `ref_name` key.

```yaml
session_templates:
- name: example-session-template
  image:
    image_ref: example-image
  configuration: example-configuration
  bos_parameters:
    boot_sets:
      example_boot_set:
        kernel_parameters: ip=dhcp quiet
        node_roles_groups:
        - Compute
        rootfs_provider: cpss3
        rootfs_provider_passthrough: dvs:api-gw-service-nmn.local:300:nmn0
```

#### HPC CSM Software Recipe variable substitutions

The HPC CSM Software Recipe provides a manifest defining the versions of each
HPC software product included in the recipe. These product versions can be used
in the `sat bootprep` input file using Jinja2 templating syntax.

> ##### Selecting the HPC CSM Software Recipe version

Author: Shouldn't it be "an HPC CSM Software Recipe version" due to the pronunciation of the HPC initialism?

    ```screen
    ncn-m001# git checkout cray/hpc-csm-software-recipe/22.11.0
    ```

1.  View the contents of the file `product_vars.yaml` in the clone of the
    repository:

    ```screen
    ncn-m001# cat product_vars.yaml
    ```

The variables defined in the `product_vars.yaml` file can be used in the values
that support templating. A variable is specified by a dot-separated path, with
each component of the path representing a key in the YAML file. For example, if
the version of the COS product appears as follows in the `product_vars.yaml`
file:

```yaml
cos:
  version: 2.4.76
```

This COS version can be used by specifying `cos.version` in a value in the
input file.

##### HPC CSM Software Recipe variable substitution example

The following example bootprep input file shows how the COS version could be
used in a bootprep input file that creates a CFS configuration for computes.
Only one layer is shown for brevity.

```yaml
---
configurations:
- name: compute-{{recipe.version}}
  layers:
  - name: cos-compute-integration-{{cos.version}}
    playbook: cos-compute.yaml
    product:
      name: cos
      version: "{{cos.version}}"
      branch: integration-{{cos.version}}
```

> Note that when the value of a key in the bootprep input file is a Jinja2

Author: The comma seems off. How about I make it a bold **Note:** like another suggestion you had?

    playbook: cos-compute.yaml
    product:
      name: cos
      version: "{{cos.version}}"
      branch: integration-{{cos.version.split('.')[0]}}-{{cos.version.split('.')[1]}}
```

#### Dynamic variable substitutions

In addition to the product version variables provided by the HPC CSM Software
Recipe as described in [HPC CSM Software Recipe variable
substitutions](#hpc-csm-software-recipe-variable-substitutions), additional
variables are available. These variables are dynamic in that their values are
determined at run-time based on the context in which they appear. The additional
variables are as follows:

- The variable `base.name` can be used in the `name` of an image in under the
  `images` key. The value of this variable is the name of the IMS image or
  recipe used as the base of this image.
- The variable `image.name` can be used in the `name` of a session template
  under the `session_templates` key. The value of this variable is the name of
  the IMS image used in this session template.

These variables reduce the need to duplicate values throughout the `sat
bootprep` input file and also make the following use cases possible:

> - Using the name of the recipe in the name of an image built from a recipe
>   provided by a product.
> - Using the name of the image in the name of a session template that references an
>   image built from a recipe provided by a product.

Reviewer: ```suggestion
- You use the name of the recipe within the name of an image, and that image is
built from a recipe provided by a product.
- You use the name of the image within the name of a session template, and that
session template references an image that is built from a recipe provided by a product.
```

Author: I think the suggestion breaks the intended meaning a little bit. I'll think about how to reword.

### Example bootprep input files

> This section provides an example bootprep input file and instructions for
> obtaining the default bootprep input files delivered with a release of the HPC
> CSM Software Recipe.

Reviewer: ```suggestion
This section provides an example bootprep input file. It also gives
instructions for obtaining the default bootprep input files delivered
with a release of the HPC CSM Software Recipe.
```

    playbook: site.yml
    product:
      name: cos
      version: "{{cos.version}}"
      branch: integration-{{cos.version}}
  - name: cpe-pe_deploy-integration-{{cpe.version}}
    playbook: pe_deploy.yml
    product:
      name: cpe
      version: "{{cpe.version}}"
      branch: integration-{{cpe.version}}

images:
- name: "{{base.name}}"
  ref_name: base_cos_image
  base:
    product:
      name: cos
      type: recipe
      version: "{{cos.version}}"

- name: compute-{{base.name}}
  ref_name: compute_image
  base:
    image_ref: base_cos_image
  configuration: compute-{{recipe.version}}
  configuration_group_names:
  - Compute

session_templates:
- name: compute-{{recipe.version}}
  image:
    image_ref: compute_image
  configuration: compute-{{recipe.version}}
  bos_parameters:
    boot_sets:
      compute:
        kernel_parameters: ip=dhcp quiet spire_join_token=${SPIRE_JOIN_TOKEN}
        node_roles_groups:
        - Compute
        rootfs_provider_passthrough: "dvs:api-gw-service-nmn.local:300:hsn0,nmn0:0"
```

#### Accessing default bootprep input files

Default bootprep input files are delivered by the HPC CSM Software Recipe

> product. These files can be accessed by cloning the `hpc-csm-software-recipe`

Reviewer: ```suggestion
product. You can access these files by cloning the `hpc-csm-software-recipe`
```

repository.

Follow steps 1-7 of the procedure in [Viewing HPC CSM Software Recipe
variables](#viewing-hpc-csm-software-recipe-variables).

> Access the files in the `bootprep` directory of that repository:

Reviewer: I would consider duplicating steps 1-7 here and adding accessing the `bootprep` directory as step 8. For new users, this file can largely be read from top to bottom. However, we'll break that flow if we don't include all the steps under this heading. With a large part of the information in this file new for Papaya, I think we want to avoid that.

Author: I really want to avoid duplicating info and having to maintain it in two places.

There's also duplicate information in the Getting Started Guide for this, and I think that will be the more visible location for this information.

Eventually, I would like to add a `sat bootprep` subcommand that outputs the variables available from the HPC CSM Software Recipe, so that other procedure would not be necessary either. See [CRAYSAT-1532](https://jira-pro.its.hpecorp.net:8443/browse/CRAYSAT-1532) for that.

Reviewer: I'm fine not duplicating the instructions until we have an idea of the repo setup with unified docs. However, I'd suggest changing the wording a little (see next comment).

> Follow steps 1-7 of the procedure in [Viewing HPC CSM Software Recipe
> variables](#viewing-hpc-csm-software-recipe-variables).
> 
> Access the files in the `bootprep` directory of that repository:

Reviewer: ```suggestion
To do this, follow steps 1-7 of the procedure in [Viewing HPC CSM Software Recipe
variables](#viewing-hpc-csm-software-recipe-variables). Then, access the files in the
`bootprep` directory of that repository:
```

```screen
ncn-m001# ls bootprep/
```

#### Generating an example bootprep input file

> Note: the `sat bootprep generate-example` command has not been updated for
> recent bootprep schema changes, and it is recommended to use the default
> bootprep input files described in [Accessing default bootprep input
> files](#accessing-default-bootprep-input-files).

Reviewer: ```suggestion
**Note:** The `sat bootprep generate-example` command was not updated for
recent bootprep schema changes. It is recommended that you instead use the
default bootprep input files described in [Accessing default bootprep input
files](#accessing-default-bootprep-input-files).
```

> 

Reviewer: I would suggest removing line 575 so all the information in this section is one paragraph.

The raw schema definition can be difficult to understand without experience

> working with JSON Schema specifications. For this reason, a feature was included
> which can generate user-friendly HTML documentation for the input file schema
> which can be browsed with the user's preferred web browser.

Reviewer: I would reword lines 612 through 614 as the following:
```suggestion
working with JSON Schema specifications. For this reason, a feature is included
that generates user-friendly HTML documentation for the input file schema
and can be browsed with your preferred web browser.
```

Author: the re-wording says "a feature ... can be browsed with your preferred web browser" which is confusing IMO. Perhaps splitting into two sentences would address your concern as well. E.g:

```
The raw schema definition can be difficult to understand without experience
working with JSON Schema specifications. For this reason, a feature is included
that generates user-friendly HTML documentation for the input file schema. This
HTML documentation can be browsed with your preferred web browser.
```
